src/sentry/middleware/auth.py

```diff
@@ -80,3 +102,43 @@ def process_exception(self, request: Request, exception):
             from sentry.web.frontend.accounts import expired
 
             return expired(request, exception.user)
+
+
+class HybridCloudAuthenticationMiddleware(MiddlewareMixin):
+    def process_request(self, request: Request):
+        from sentry.web.frontend.accounts import expired
+
+        auth_result = auth_service.authenticate(request=authentication_request_from(request))
+        request.user_from_signed_request = auth_result.user_from_signed_request
+
+        if auth_result.auth is not None:
+            request.auth = auth_result.auth
+        if auth_result.expired:
+            return expired(request, auth_result.user)
+        elif auth_result.user is not None:
+            request.user = auth_result.user
+            UserIP.log(auth_result.user, request.META["REMOTE_ADDR"])
+        else:
+            request.user = AnonymousUser()
+
+
+def authentication_request_from(request: Request) -> AuthenticationRequest:
```

Comment:
Should this be a method on the auth_service, or in the auth_service module?

Reply:
Yes, in a follow up PR when I apply a similar approach to the `authentication_classes` of our json `Endpoints` I will do this.  Right now it was sensible to keep it local here.

Reply:
Sounds good to me.